🚀 3단계 - 테스트를 통한 코드 보호

writer0713 opened the pull request:
## 5.15

안녕하세요, 멘토님.
`@SpringBootTest` 를 사용한 테스트 코드 작성했습니다.
아직 다 작성하지는 못했고 디펜던시가 비교적 적은 `메뉴그룹`, `테이블`, `상품` 세가지에 대한 테스트코드 먼저 작성하여 리뷰 요청 드립니다.
이번 리뷰도 잘 부탁 드립니다 :) 

## 5.18
안녕하세요. 멘토님.
피드백 주신 부분들 수정 및 반영했습니다.
그리고, 추가로 `메뉴` 관련 테스트 케이스 작성 했습니다.

이전 피드백에서 알려주신 `테스트 더블` 관련된 포스팅 보면서 여러가지 생각을 하고 있는데요.
우선은 `@SpringbootTest` + `Mockito` 를 사용한 방식으로 작성해보고,
그 이후에 다른 기법으로 리팩토링을 해볼까 합니다 :) 

## 5.22
- 남은 테스트 (주문 도메인) 추가 했습니다
- 이전에 주신 피드백 적용했습니다
  - 경계값 사용
  - fixture 사용

그 외,

> 테스트 더블에 대해 알아보고, mock 외에도 여러가지 기법을 이용해보시면 더욱 재밌을 것 같아요 😄

요 부분은 자료를 좀 더 찾아보고 작성해보겠습니다 :) 

> mock 을 이용하는 것과 아닌 것에는 어떠한 장단점이 있을까요? 😉

요 질문도 좀 더 자료 찾아보고 실제 mock 을 사용하지 않고 다른 방식으로 테스트 구현해본 이후에 느낀점 답변 드리도록 하겠습니다 :)

writer0713 committed:
docs: step-0 에서 진행해야 하는 todos 를 README.md 에 정리
- README.md 에 todos 정리

writer0713 committed:
test: CarTest 추가
- CarTest 추가
- junit test method 추가

writer0713 committed:
docs: step0 todos README.md 수정
- check 가능하도록 md 수정

writer0713 committed:
feat: Car 클래스 추가
- Car 클래스 추가

writer0713 committed:
test: 자동차 이름 관련 테스트 추가
- 자동차 이름은 5글자를 넘을수 없다.
- 5글자가 넘으면 IllegalArgumentException 이 발생한다.

writer0713 committed:
feat: Car 클래스 name length 제한 추가
- 5 글자 넘으면 throw IllegalArgumentException

writer0713 committed:
feat: Car 의 move 메서드 구현
- move 메서드 구현
- 0~9 사이의 숫자만 들어갈수 있도록 validation
- 4 이상의 숫자가 들어왔을때에만 currentPosition 변경

writer0713 committed:
test: Car 의 move 메서드 테스트 추가
- move 메서드의 range validation test
- move 메서드에 4 보다 작은값 넣은 경우 test
- move 메서드에 4 이상 넣은 경우 test

writer0713 committed:
refactor: Car 의 move 메서드 수정
- position 대신 condition 을 받도록 수정
- condition 은 0~9 사이의 값을 받되, 4 이상의 값인 경우에만 position 에 1을 더하도록 수정

writer0713 committed:
test: Car 의 move 메서드 수정에 따른 테스트 수정
- car 의 move 메서드 수정에 따라 테스트 expectation 값 수정

writer0713 committed:
feat: Car 의 move 메서드 수정
- move 메서드가 condition 에 따라 움직일수 있는지 판단하는 책임을 MovingStrategy 에 맡기도록 수정
- 외부에 공개된 move 메서드는 condition 을 받아서 0~9 인지 validation
- 실제 움직일수 있는지는 private move 메서드를 사용

writer0713 committed:
docs: README.md 수정
- 다 진행한 todo 는 체크

writer0713 committed:
refactor: MovingStrategy 구현체 추가
- MovingStrategy 의 구현체인 MovingStrategy 를 추가
- MovableStrategy 사용하도록해서 move 메서드의 indent 를 한칸으로 줄임

writer0713 committed:
Merge branch 'next-step:writer0713' into writer0713

writer0713 committed:
Merge branch 'next-step:writer0713' into writer0713

writer0713 committed:
Merge branch 'next-step:writer0713' into writer0713

writer0713 committed:
docs: 요구 사항 정리 - 테이블 관련 표현을 하나로 통일 (step-2)
- '테이블을 정리하다' 라는 표현으로 통일

writer0713 committed:
docs: 요구 사항 정리 - 상품에 대한 예시 추가 (step-2)
- 상품에 대한 예시 추가

writer0713 committed:
docs: 요구 사항 정리 - 메뉴 그룹에 대한 예시 추가 (step-2)
- 메뉴 그룹에 대한 예시 추가

writer0713 committed:
docs: 요구 사항 정리 - 메뉴에 대한 예시 추가 (step-2)
- 메뉴에 대한 예시 추가

writer0713 committed:
docs: 요구 사항 정리 - 주문에 대한 예시 추가 (step-2)
- 주문에 대한 예시 추가

writer0713 committed:
docs: 요구 사항 정리 - 스타일 일치 (step-2)
- 예시는 요구사항 다음줄에 위치시키도록 스타일 수정

writer0713 committed:
test: 메뉴그룹 생성 관련 테스트 추가
- 메뉴 그룹 등록이 가능하다.
- 메뉴 그룹 이름을 꼭 입력해야 한다.

writer0713 committed:
test: @SpringBootTest 를 위한 data init 세팅
- 테스트용 data.sql 추가 (data init)
- jpa 로 schema 가 만들어진 이후에 data.sql 이 실행되도록 설정 추가

writer0713 committed:
test: 메뉴그룹 목록 조회 테스트 추가
- 등록된 메뉴 그룹 목록을 조회할 수 있다

writer0713 committed:
test: product 상품 목록 조회 테스트 추가
- 등록된 상품 목록을 조회할 수 있다.

writer0713 committed:
test: 테스트 함수명 수정
- 동사_명사_test 와 같은 함수명으로 수정

writer0713 committed:
test: 상품 등록 관련 테스트 추가
- 상품 등록이 가능하다.
- 상품 이름에는 비속어가 들어갈 수 없다.
- 상품 가격을 꼭 입력해야 한다.
- 가격은 음수일수 없다.

writer0713 committed:
test: 메뉴그룹 목록 조회 테스트 수정
- 다른 테스트에 영향을 받지 않도록 조정

writer0713 committed:
feat: '-' 가 없는 UUID (String) 을 UUID 로 바꿔주는 유틸 추가
- ex) "3b52824434f7406bbb7e690912f66b10"

writer0713 committed:
test: 등록된 상품의 가격 수정 관련 테스트 추가
- 등록된 상품의 가격을 수정할 수 있다.
- 가격은 음수일 수 없다.
- 해당 상품으로 구성된 메뉴의 가격이 수정된 상품의 가격 총합보다 크면 메뉴를 비노출한다.

writer0713 committed:
test: @Nested 사용하여 관련된 테스트들 묶도록 수정

writer0713 committed:
test: OrderTable 관련 테스트 추가
- 사장님은 가게의 테이블에 이름을 부여하여 등록할 수 있다.
- 최초 테이블의 상태는 다음과 같이 등록된다.
- 손님이 테이블에 앉으면 테이블의 상태는 다음과 같이 변경된다.
- 테이블을 사용하는 손님의 숫자를 변경시에는 0보다 큰 숫자만 가능하다.
- 테이블을 사용하는 손님이 있을때에만 사용하는 손님의 숫자를 변경 가능하다.
- 손님이 테이블을 사용 완료하여 테이블을 정리하면 다시 상태를 변경할 수 있다.
- 주문 상태가 `완료` 인 경우에만, 해당 주문 테이블을 정리할 수 있다.
- 등록된 테이블의 목록을 조회할 수 있다.

Hyeon9mak commented on src/test/java/kitchenpos/application/MenuGroupServiceTest.kt:
```diff
@@ -0,0 +1,58 @@
+    private val sut: MenuGroupService,
+    private val menuGroupRepository: MenuGroupRepository,
+) {
+    @DisplayName("메뉴 그룹 등록이 가능하다.")
+    @Nested
```
`@Nested` 활용 👍

Hyeon9mak commented on src/test/java/kitchenpos/application/MenuGroupServiceTest.kt:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    @DisplayName("등록된 메뉴그룹 목록을 조회할 수 있다")
+    @Test
+    fun find_all_menu_group_test() {
+        // given
+        // when
+        val menuGroups = sut.findAll()
+
+        // then
+        assertThat(menuGroups.size).isGreaterThan(0)
+    }
```
하나의 DB 를 이용해서 여러가지 테스트를 진행하는 경우 서로 각기의 테스트에 영향을 받을텐데요,
이러면 테스트를 추가하거나 유지보수 할 때 계속 불편함을 겪게 될거에요.

`BeforeEach` 혹은 `AfterEach` 를 이용하여 데이터를 초기화하거나 셋업하여 테스트를 진행해보는 것은 어떨까요? 😉 
https://www.baeldung.com/junit-before-beforeclass-beforeeach-beforeall

writer0713 replied:
`@Sql` 어노테이션과 `data.sql` 을 사용하여 각 테스트 실행전에 데이터를 초기화하도록 수정하겠습니다 :) 
참고 : https://yenjjun187.tistory.com/797

writer0713 committed:
test: ProductServiceTest 잘못된 변수명 수정
- purgomalumClonneable => purgomalumClient

Hyeon9mak commented on src/test/java/kitchenpos/application/OrderTableServiceTest.kt:
```diff
@@ -0,0 +1,171 @@
+            assertThat(createdTable.id).isNotNull()
+        }
+
+        @DisplayName("최초 테이블의 상태는 다음과 같이 등록된다.")
+        @Test
+        fun case_2() {
+            // given
+            val request = OrderTable()
+            request.name = "1번 테이블"
+
+            // when
+            val createdTable = sut.create(request)
+
+            // then
+            with(createdTable) {
+                assertThat(this.id).isNotNull()
+                assertThat(this.numberOfGuests).isEqualTo(0)
+                assertThat(this.isOccupied).isFalse()
+            }
+        }
+    }
```
테스트 명칭과 실제 검증 내용이 조금 상이한 것 같아요 😱

writer0713 replied:
@Hyeon9mak 
최초 테이블 등록시 테이블 상태 (손님수 0명 / 빈자리) 에 대한 테스트인데요.
테스트 명칭이 조금 모호한것 같아 수정하겠습니다.

Hyeon9mak commented on src/test/java/kitchenpos/application/OrderTableServiceTest.kt:
```diff
@@ -0,0 +1,171 @@
+import org.springframework.boot.test.mock.mockito.MockBean
+import org.springframework.test.context.TestConstructor
+
+@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
+@SpringBootTest
+class OrderTableServiceTest {
+    @MockBean
+    private lateinit var orderRepository: OrderRepository
+
+    @Autowired
+    private lateinit var sut: OrderTableService
+
+    @DisplayName("사장님은 가게의 테이블을 등록할 수 있다.")
```
mock 을 이용하는 것과 아닌 것에는 어떠한 장단점이 있을까요? 😉

Hyeon9mak commented on src/test/java/kitchenpos/application/ProductServiceTest.kt:
```diff
@@ -0,0 +1,160 @@
+import org.springframework.boot.test.mock.mockito.MockBean
+import java.math.BigDecimal
+
+@SpringBootTest
+class ProductServiceTest {
+    @MockBean
+    private lateinit var purgomalumCloneable: PurgomalumClient
```
https://hudi.blog/test-double/
테스트 더블에 대해 알아보고, mock 외에도 여러가지 기법을 이용해보시면 더욱 재밌을 것 같아요 😄

Hyeon9mak reviewed (changes requested):
안녕하세요 정환님! 테스트 코드 탄탄하게 잘 작성해주셨네요!
나머지 테스트들도 기대되는데요? 😄 👍 
몇 가지 코멘트 남겼습니다!

writer0713 committed:
test: OrderTableServiceTest 일부 테스트 @DisplayName 수정
- 좀 더 구체적으로 수정

writer0713 committed:
docs: README.md 에서 메뉴에 대한 요구사항 수정
- 메뉴에 대한 요구사항 추가/수정

writer0713 committed:
test: MenuServiceTest 추가
- 등록된 상품 목록을 조회할 수 있다.
- 메뉴의 가격은 `0원` 보다 작을 수
- 메뉴는 이미 등록된 `메뉴 그룹` 에 속한다.

writer0713 committed:
test: 테스트용 데이터 init 을 위한 data.sql 수정
- @Sql 어노테이션을 사용하여 각 테스트케이스 직전에 데이터를 초기화 하기 위함
- 우선 테이블에 있는 데이터를 모두 삭제하고 다시 모든 데이터를 추가하도록 수정

writer0713 committed:
test: @Sql 어노테이션 사용하도록 수정
- @Sql 어노테이션 사용하여 'db/data.sql' 사용하여 각 테스트케이스 실행전에 데이터 초기화 하도록 수정
- 'isGreaterThan' 사용하던 부분을 'isEqualTo' 로 바꿔서 더 정확한 테스트가 되도록 수정

writer0713 committed:
test:MenuServiceTest 메뉴등록 관련테스트케이스 추가
- 메뉴의 가격은 `0원` 보다 작을 수 없다.
- 메뉴는 이미 등록된 `메뉴 그룹` 에 속한다.
- 메뉴는 1개 이상의 `메뉴 구성 상품` 을 가지고 있어야 한다.
- 메뉴의 `메뉴 구성 상품` 은 이미 등록된 `상품` 으로만 등록 가능하다.
- `메뉴 구성 상품` 의 개수와 `상품` 의 개수는 같아야 한다.
- `메뉴 구성 상품` 의 수량은 `0개` 보다 작을수 없다.
- 메뉴의 가격은 `메뉴 구성 상품` 가격의 총합보다 클 수 없다.
- 메뉴 이름에는 비속어가 들어갈 수 없다.
- 메뉴는 이름, 가격, 메뉴 구성 상품 정보를 꼭 입력해야 한다.

writer0713 committed:
test: MenuServiceTest 메뉴 노출상태 변경 관련 테스트 작성
- 메뉴를 숨길수 있다.
- 메뉴를 노출할 수 있다.
- 메뉴의 가격이 `메뉴 구성 상품` 가격 총합보다 크면 노출할 수 없다.

writer0713 committed:
test: MenuServiceTest 에 메뉴 가격 수정 관련 테스트 추가
- 메뉴의 가격은 `0원` 보다 작을 수 없다.
- 메뉴의 가격은 `메뉴 구성 상품` 가격의 총합보다 클 수 없다.

writer0713 commented:
@Hyeon9mak 

안녕하세요. 멘토님.
피드백 주신 부분들 수정 및 반영했습니다.
그리고, 추가로 메뉴 관련 테스트 케이스 작성 했습니다.

이전 피드백에서 알려주신 테스트 더블 관련된 포스팅 보면서 여러가지 생각을 하고 있는데요.
우선은 @SpringbootTest + Mockito 를 사용한 방식으로 작성해보고,
그 이후에 다른 기법으로 리팩토링을 해볼까 합니다 :)

Hyeon9mak commented on src/test/java/kitchenpos/application/MenuServiceTest.kt:
```diff
@@ -0,0 +1,320 @@
+import java.util.*
+
+@SpringBootTest
+@Sql("classpath:db/data.sql")
```
`@Sql` 을 사용하는 것도 좋은 방법이죠! 👍

Hyeon9mak commented on src/test/resources/application.properties:
```diff
@@ -3,4 +3,5 @@ spring.datasource.username=sa
 spring.flyway.enabled=false
 spring.jpa.properties.hibernate.format_sql=true
 spring.jpa.show-sql=true
+spring.jpa.defer-datasource-initialization=true
```
👍 👍

Hyeon9mak commented on src/test/java/kitchenpos/application/MenuServiceTest.kt:
```diff
@@ -0,0 +1,320 @@
+    @DisplayName("메뉴를 등록할 수 있다.")
+    @Nested
+    inner class CreateMenu {
+        @DisplayName("메뉴의 가격은 `0원` 보다 작을 수 없다.")
+        @Test
+        fun case_1() {
+            // given
+            val request = Menu()
+            request.price = BigDecimal.valueOf(-1000)
+
+            // when
+            // then
+            assertThrows<IllegalArgumentException> { sut.create(request) }
+        }
```
경계값을 이용하여 테스트해준다면 아~~주 조금 더 확실한 검증이 될거 같아요 😄 👍 

```java
request.price = BigDecimal.valueOf(-1)
```

Hyeon9mak replied:
요기는 수정이 살짝 누락된거 같네요 😉

Hyeon9mak commented on src/test/java/kitchenpos/application/MenuServiceTest.kt:
```diff
@@ -0,0 +1,320 @@
+            assertThrows<IllegalArgumentException> { sut.create(request) }
+        }
+
+        @DisplayName("메뉴의 `메뉴 구성 상품` 은 이미 등록된 `상품` 으로만 등록 가능하다. (w/ 존재하는 상품으로 등록)")
+        @Test
+        fun case_5() {
+            // given
+            val menuProduct = generateMenuProduct(productId = EXISTING_PRODUCT_ID_1)
+
+            val request = Menu()
+            request.price = BigDecimal.valueOf(16_000)
+            request.menuGroupId = EXISTING_MENU_GROUP_ID
+            request.menuProducts = listOf(menuProduct)
+            request.name = "후라이드 치킨"
+
+            // when
+            val createdMenu = sut.create(request)
+
+            // then
+            assertThat(createdMenu.id).isNotNull()
```
굉장히 다양한 케이스에 대해서도 모두 다 테스트를 작성해주셨네요 😮 👍 👍

Hyeon9mak commented on src/test/java/kitchenpos/application/MenuServiceTest.kt:
```diff
@@ -0,0 +1,320 @@
+            // then
+            assertThrows<NoSuchElementException> { sut.create(request) }
+        }
+
+        @DisplayName("메뉴는 1개 이상의 `메뉴 구성 상품` 을 가지고 있어야 한다. ")
+        @Test
+        fun case_3() {
+            // given
+            val request = Menu()
+            request.price = BigDecimal.valueOf(16_000)
+            request.menuGroupId = EXISTING_MENU_GROUP_ID
+            request.menuProducts = null
+
+            // when
+            // then
+            assertThrows<IllegalArgumentException> { sut.create(request) }
+        }
+
+        @DisplayName("메뉴의 `메뉴 구성 상품` 은 이미 등록된 `상품` 으로만 등록 가능하다. (w/ 존재하지 않는 상품으로 등록)")
+        @Test
+        fun case_4() {
+            // given
+            val request = Menu()
+            request.price = BigDecimal.valueOf(16_000)
+            request.menuGroupId = EXISTING_MENU_GROUP_ID
+            request.menuProducts = null
+
```
계속해서 반복되어 준비되는 셋팅 값들이 있고,
걔중에 테스틀 위해서만 1 개의 값만 살짝 변경되는 중인데요,
이런 경우 이미 fixture method 를 만들어둔다면 테스트 코드 작성에도 편의를 얻고,
테스트를 읽는 입장에서도 변경된 값에 대해서만 집중할 수 있을 것 같아요 😄 

```kotlin

fun initMenu(
  price: BigDecimal = BigDecimal.valueOf(16_000),
  menuGroupId: Long = EXISTING_MENU_GROUP_ID
  menuProducts: List<MenuProduct> = listOf(...)
): Menu {
  return Menu(...)
}
```

kotlin 에서는 메서드의 기본 파라미터 기능을 제공하므로, 이를 더욱 쉽게 준비할 수 있겠네요!

Hyeon9mak reviewed (changes requested):
안녕하세요 정환님! 나머지 테스트 코드도 잘 작성해주셨습니다 👍 
워낙 꼼꼼하게 케이스별로 잘 작성해주셔서 사실 드릴 코멘트가 많지 않은데요 😄 
테스트 코드 작성 편의를 위해 kotlin 이 제공하는 기능을 이용해 fixture 코드를 작성해보시는 건 어떨까 코멘트 남겼습니다.
그 외에도 mock 을 이용하시며 느끼신 장단점이 궁금하네요~!
궁금하신 점 언제든지 코멘트 남겨주세요!

writer0713 committed:
test: 주문 관련 테스트 케이스 작성
- data.sql 에서 메뉴 한개 비노출 상태로 수정 (비노출 관련 테스트 위함)

writer0713 committed:
test: fixture 추가
- fixture 추가 및 테스트 코드에서 fixture 사용하도록 수정

writer0713 committed:
test: 경계값을 이용하도록 테스트 수정

writer0713 commented:
@Hyeon9mak 
- 남은 테스트 (주문 도메인) 추가 했습니다
- 이전에 주신 피드백 적용했습니다
  - 경계값 사용
  - fixture 사용

그 외,

> 테스트 더블에 대해 알아보고, mock 외에도 여러가지 기법을 이용해보시면 더욱 재밌을 것 같아요 😄

요 부분은 자료를 좀 더 찾아보고 작성해보겠습니다 :) 

> mock 을 이용하는 것과 아닌 것에는 어떠한 장단점이 있을까요? 😉

요 질문도 좀 더 자료 찾아보고 실제 mock 을 사용하지 않고 다른 방식으로 테스트 구현해본 이후에 느낀점 답변 드리도록 하겠습니다 :)

Hyeon9mak commented on src/test/java/kitchenpos/fixture/Fixtures.kt:
```diff
@@ -0,0 +1,127 @@
+import java.util.UUID
+
+// menu-group
+val EXISTING_MENU_GROUP_ID = generateUUIDFrom("f1860abc2ea1411bbd4abaa44f0d5580")
+
+// menu
+val EXISTING_MENU_ID = generateUUIDFrom("b9c670b04ef5409083496868df1c7d62")
+val NOT_EXISTING_MENU_GROUP_ID = generateUUIDFrom("f1860abc2ea1411bbd4abaa44f0d1111")
+val NON_DISPLAYED_MENU_ID = generateUUIDFrom("33e558df7d934622b50efcc4282cd184")
+
+// order-table
+val EXISTING_ORDER_TABLE_ID = generateUUIDFrom("8d71004329b6420e8452233f5a035520")
+
+// product
+val EXISTING_PRODUCT_ID_1 = generateUUIDFrom("3b52824434f7406bbb7e690912f66b10")
+val EXISTING_PRODUCT_ID_2 = generateUUIDFrom("c5ee925c3dbb4941b825021446f24446")
```
fixture 생성 👍 👍

Hyeon9mak commented on src/test/java/kitchenpos/application/OrderTableServiceTest.kt:
```diff
@@ -0,0 +1,168 @@
+    @Nested
+    inner class Sit {
+        @DisplayName("테이블을 사용하는 손님이 없을때에는 사용하는 손님의 숫자를 변경할 수 없다.")
+        @Test
+        fun case_4() {
+            // given
+            val tableId = createTable().id
+            val request = initOrderTable(id = tableId, numberOfGuests = 3)
+
+            // when
+            // then
+            assertThrows<IllegalStateException> { sut.changeNumberOfGuests(tableId, request) }
```
크크크 픽스쳐 덕분에 테스트 하고자 하는 셋팅이 무엇인지 더 명확히 눈에 들어오네요!!

Hyeon9mak reviewed (approved):
안녕하세요 정환님! fixture 생성도 잘 해주셨습니다 👍 👍 
덕분에 테스트 세팅값을 통해 "무엇을 테스트 하고 싶은지" 가 한 눈에 들어오네요 😄 👍 

요구사항도 충족된 것 같고, 피드백도 잘 반영해주셔서 approve 하고자 합니다!
그동안 남겨드린 코멘트들이 도움이 되셨을까요?

남은 미션들도 재밌으니 끝까지 달려보시길 바랍니다!
(mock 관련 장단점은 나중에라도 느끼신 점 알려주세요 😄 👍 )
고생 많으셨습니다!

writer0713 commented:
@Hyeon9mak 
꼼꼼하게 리뷰해주셔서 감사합니다 :)
